io/EptReader.cpp

```diff
@@ -902,86 +893,120 @@ Dimension::Type EptReader::getCoercedTypeTest(const NL::json& j)
-    m_currentIndex = 0;
+void EptReader::load()
+{
+    // Asynchronously trigger the fetching and point-view execution of
+    // a lookahead buffer of nodes.
+    while (
+        m_upcomingNodeBuffers.size() < m_pool->size() &&
+        m_overlapIt != m_overlaps.end())
+    {
+        const auto nodeId(m_nodeId++);
+        const auto key(m_overlapIt->first);
+        ++m_overlapIt;
+
+        log()->get(LogLevel::Debug) << nodeId << "/" << m_overlaps.size() <<
+            std::endl;
+
+        // Insert an empty placeholder node to keep track of the outstanding
+        // nodes that are currently being fetched/executed.
+        std::unique_lock<std::mutex> lock(m_mutex);
+        auto& loadingBuffer = m_upcomingNodeBuffers[nodeId];
+        lock.unlock();
+
+        m_pool->add([this, &loadingBuffer, nodeId, key]()
+        {
+            std::unique_ptr<NodeBuffer> nodeBuffer(
+                new NodeBuffer(*m_userLayout));
 
-    // Read addon information after the native data, we'll possibly
-    // overwrite attributes.
-    for (const auto& addon : m_addons)
-        readAddon(*m_bufferPointView, key, *addon, startId);
+            if (m_info->dataType() == EptInfo::DataType::Laszip)
+                readLaszip(nodeBuffer->view, key, nodeId);
+            else
+                readBinary(nodeBuffer->view, key, nodeId);
 
-    m_nodeId++;
-}
+            for (const auto& addon : m_addons)
+                readAddon(nodeBuffer->view, key, *addon);
+
+            std::unique_lock<std::mutex> lock(m_mutex);
+            loadingBuffer = std::move(nodeBuffer);
+            lock.unlock();
 
+            // A node has been populated - notify our consumer thread.
+            m_cv.notify_one();
+        });
+    }
+}
 
-void EptReader::fillPoint(PointRef& point)
+EptReader::NodeBufferIt EptReader::findBuffer()
 {
-    DimTypeList dims = m_bufferPointView->dimTypes();
-    m_bufferPointView->getPackedPoint(dims, m_currentIndex,
-        m_temp_buffer.data());
-    point.setPackedData(dims, m_temp_buffer.data());
-    m_currentIndex++;
+    return std::find_if(
+        m_upcomingNodeBuffers.begin(),
+        m_upcomingNodeBuffers.end(),
+        [this](const NodeBufferPair& p) { return !!p.second; });
 }
 
+bool EptReader::next()
+{
+    // Asynchronously trigger the loading of some nodes.
+    load();
+
+    // Now wait for a node to be populated, or for there to be no outstanding
+    // nodes left, in which case we're all done.
+    std::unique_lock<std::mutex> lock(m_mutex);
+    m_cv.wait(lock, [this]()
+    {
+        return m_upcomingNodeBuffers.empty() ||
+            findBuffer() != m_upcomingNodeBuffers.end();
+    });
+
+    // Processing is complete.
+    if (m_upcomingNodeBuffers.empty()) return false;
+
+    // We know at least one node is populated from our `wait` predicate, so find
+    // the first one and transfer it out of our `upcoming` pool to make it the
+    // current active node.
+    const auto it = findBuffer();
```

Comment:
The result of this iterator should be sufficient.

Reply:
Resolved in 807b0be20.